src/ext/java/org/opentripplanner/ext/siri/SiriTripPatternIdGenerator.java

```diff
@@ -0,0 +1,28 @@
+package org.opentripplanner.ext.siri;
+
+import org.opentripplanner.gtfs.GenerateTripPatternsOperation;
+import org.opentripplanner.model.FeedScopedId;
+import org.opentripplanner.model.Route;
+import org.opentripplanner.model.Trip;
+
+class SiriTripPatternIdGenerator {
+  private int counter = 0;
+
+  /**
+   * Generate unique trip pattern code for real-time added trip pattern. This function roughly
+   * follows the format of {@link GenerateTripPatternsOperation#generateUniqueIdForTripPattern(Route, int)}.
+   * <p>
+   * The generator add a postfix 'RT' to indicate that this trip pattern is generated at REAL-TIME.
+   */
+  FeedScopedId generateUniqueTripPatternId(Trip trip) {
+    Route route = trip.getRoute();
+    FeedScopedId routeId = route.getId();
+    String directionId = trip.getDirectionId();
+    if( directionId == null) { directionId = ""; }
+
+    // OBA library uses underscore as separator, we're moving toward colon.
+    String id = String.format("%s:%s:%03d:RT", routeId.getId(), directionId, ++counter);
```

Comment:
It would be good to have a comment explaining why unsynchronized increment-and-read is not problematic here. I see that this is called from one synchronized method, but also from one other unsynchronized method and it's not immediately obvious whether they can both run concurrently on different threads.

Or to avoid placing lots of special conditions on use of this method it could just be synchronized (which has very low cost when there's no contention).